bin/GoogleAdsCleanupServices.php

```diff
@@ -0,0 +1,410 @@
+	}
+
+	/**
+	 * Find used enums and remove any unused ones.
+	 */
+	protected function remove_enums() {
+		$this->output_text( 'Removing unused enums from Google Ads library ' . $this->version );
+
+		$library_enums = $this->find_library_file_pattern(
+			"{$this->path}/metadata/Google/Ads/GoogleAds/{$this->version}/Enums"
+		);
+
+		$used_enums = $this->find_used_pattern(
+			"use Google\\\\Ads\\\\GoogleAds\\\\{$this->version}\\\\Enums\\\\([A-Za-z0-9]+)Enum\\\\"
+		);
+
+		foreach ( array_diff( $library_enums, $used_enums ) as $enum ) {
+			$this->remove_enum( $enum );
+		};
+	}
+
+	/**
+	 * Remove a specific enum.
+	 *
+	 * @param string $enum Enum name.
+	 */
+	protected function remove_enum( string $enum ) {
+		$this->output_text( "Removing enum {$enum}" );
+		$this->src_path = '/src/Google/Ads/GoogleAds';
+
+		$this->remove_file( "/{$this->version}/Enums/{$enum}Enum.php" );
+		$this->remove_file( "/{$this->version}/Enums/{$enum}Enum_{$enum}.php" );
+		$this->remove_file( "/{$this->version}/Enums/{$enum}Enum/{$enum}.php" );
+		$this->remove_directory( "/{$this->version}/Enums/{$enum}Enum" );
+
+		// Remove metadata files.
+		$this->src_path = '/metadata/Google/Ads/GoogleAds';
+		$this->remove_file( "/{$this->version}/Enums/{$enum}.php" );
+	}
+
+	/**
+	 * Find a specific pattern used within the extension.
+	 *
+	 * @param string $pattern Regexp pattern to match.
+	 *
+	 * @return array List of names that match.
+	 */
+	protected function find_used_pattern( string $pattern ): array {
+		$command = "grep -rE --include=*.php '{$pattern}' {$this->code_path}";
+
+		// phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions.system_calls_exec
+		exec( $command, $output );
+
+		if ( empty( $output ) ) {
+			return [];
+		}
+
+		return array_unique(
+			array_map(
+				function( $line ) use ( $pattern ) {
+					preg_match( "/{$pattern}/", $line, $matches );
+					return $matches[1];
+				},
+				$output
+			)
+		);
+	}
+
+	/**
+	 * Find a specific filename pattern within the library.
+	 *
+	 * @param string $pattern Regexp pattern to match.
+	 * @param string $suffix  Suffix to remove from filename.
+	 *
+	 * @return array List of matched names.
+	 */
+	protected function find_library_file_pattern( string $pattern, string $suffix = null ): array {
+		$output = glob( "{$pattern}/*.php" );
+
+		if ( empty( $output ) ) {
+			return [];
+		}
+
+		return array_map(
+			function( $file ) use ( $suffix ) {
+				$name = pathinfo( $file, PATHINFO_FILENAME );
+				return $suffix ? $this->remove_suffix( $suffix, $name ) : $name;
+			},
+			$output
+		);
+	}
+
+	/**
+	 * Optionally remove a suffix from a string.
+	 *
+	 * @param string $suffix Suffix to remove.
+	 * @param string $text   Text to remove the suffix from.
+	 *
+	 * @return string
+	 */
+	protected function remove_suffix( string $suffix, string $text ): string {
+		return preg_replace( '/' . preg_quote( $suffix, '/' ) . '$/', '', $text );
```

Comment:
❓ 

Why do we need to escape  '/' in the `preg_quote`?  As far as I know `pathinfo( $file, PATHINFO_FILENAME )` will return just the file name with no `/` cases.

Reply:
We aren't escaping `/`, we are passing that as the delimiter character. We are escaping any [special character](https://www.php.net/manual/en/function.preg-quote.php).

Not strictly necessary for our use case since the suffix contains only letters. But since it's a function called `remove_suffix` it could be used for other strings. And you could also argue that `PATHINFO_FILENAME` could in some cases return a string with a `.` which is on the list of special characters.